activerecord/lib/active_record/connection_adapters/abstract/query_comment.rb

```diff
@@ -0,0 +1,178 @@
+        mattr_accessor :cache_query_comment, instance_accessor: false, default: true
+        mattr_accessor :backtrace_cleaner, default: ActiveSupport::BacktraceCleaner.new
+        thread_mattr_accessor :cached_comment, instance_accessor: false
+
+        class << self
+          # Updates the context used to construct the query comment.
+          # Resets the cached comment if <tt>cache_query_comment</tt> is +true+.
+          def update(ctx)
+            return unless ctx.is_a? Hash
+            self.context.merge!(ctx.symbolize_keys)
+            self.cached_comment = nil
+          end
+
+          def comments_available?
+            self.components.present? || self.inline_annotations.present?
+          end
+
+          # Returns a +String+ containing the component-based query comment.
+          # Sets and returns a cached comment if <tt>cache_query_comment</tt> is +true+.
+          def comment
+            return cached_comment if cache_query_comment && !cached_comment.nil?
+            self.cached_comment = "/*#{escape_sql_comment(comment_content)}*/"
+          end
+
+          # Returns a +String+ containing any inline comments from +with_annotation+.
+          def inline_comment
+            return nil unless inline_annotations.present?
+            "/*#{escape_sql_comment(inline_comment_content)}*/"
+          end
+
+          # Manually clear the comment cache.
+          def clear_comment_cache!
+            self.cached_comment = nil
+          end
+
+          # Annotate any query within `&block`. Can be nested.
+          def with_annotation(comment, &block)
+            self.inline_annotations.push(comment)
+            block.call if block.present?
+          ensure
+            self.inline_annotations.pop
+          end
+
+          # Return the set of active inline annotations from +with_annotation+.
+          def inline_annotations
+            context[:inline_annotations] ||= []
+          end
+
+          # QueryComment +component+ methods
+
+          # Set during Rails boot in lib/active_record/railtie.rb
+          def application # :nodoc:
+            context[:application_name] || "active_record"
+          end
+
+          def pid # :nodoc:
+            Process.pid
+          end
+
+          def connection_config # :nodoc:
+            ActiveRecord::Base.connection_db_config
+          end
+
+          def socket # :nodoc:
+            connection_config.socket
+          end
+
+          def db_host # :nodoc:
+            connection_config.host
+          end
+
+          def database # :nodoc:
+            connection_config.database
+          end
+
+          def line # :nodoc:
+            backtrace_cleaner.add_silencer { |line| line.match?(/lib\/active_(record|support)/) }
+            backtrace_cleaner.clean(caller.lazy).first
+          end
+
+          private
+
+          def context
+            Thread.current[:active_record_query_comment_context] ||= {}
+          end
```

Comment:
This is what I landed on to solve the nested thread problem.

Alternatives I considered:

* Handling `nil`s wherever `context` was referenced directly (safe navigation).
* Wrapping `context` in another method to check for existence and referring to that method internally.

This feels like a good compromise.